### PR TITLE
[tt-train] Add gradient norm clipping

### PR DESCRIPTION
**Pull request by jaykru-tt**
### What's changed
- Introduces on-device gradient clipping for tt-train using Moreh's on-device gradient clipping op.
- Introduces a test with a C++ reference implementation of gradient clipping over floats -> tolerances are currently high, likely due to bf16 imprecision that we don't currently have a way to emulate in host-side C++ code. Due to these high tolerances, I've annotated the test with GENEROUS_TOLERANCE so we don't forget to clean up the test once we add arithmetic operations to metal's host-side bf16 class.

### Checklist
- [x] New test passes

### Evaluation of changes
image
Key: 

- Black: main/baseline
- Red: with change, NOT using clip grad norm: identical to main.
- Green: with changes, using clip grad norm: moderate improvement in loss curve early on, trading blows toward end of training.

### CODE REVIEW

File: tt-train/sources/ttml/core/clip_grad_norm.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    const serialization::NamedParameters& parameters, float max_norm, float p_norm_type, bool error_if_nonfinite) {
+    std::vector<tt::tt_metal::Tensor> grads;
+    grads.reserve(parameters.size());
+    for (const auto& [_, tensor] : parameters) {
+        if (tensor->get_requires_grad() && tensor->is_grad_initialized()) {
+            grads.push_back(tensor->get_grad());
+        }
+    }
+    auto tt_result = ttnn::moreh_clip_grad_norm(
+        grads, max_norm, p_norm_type, error_if_nonfinite, std::nullopt, std::nullopt, std::nullopt);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.cpp:22]: add comments to std::nullopt

example:
```
/* name of var */ std::nullopt, ...
```

File: tt-train/sources/ttml/core/clip_grad_norm.hpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#include "serialization/serializable.hpp"
+
+namespace ttml::core {
+
+// Clip the gradients of the parameters up to a given maximum norm. If
+// error_if_nonfinite is true, an error is thrown if the sum of the parameters
+// is in {nan,inf,-inf}. p_norm_type specifies which p-norm
+// (https://en.wikipedia.org/wiki/Norm_(mathematics)#p-norm) to use in the norm
+// calculation. Gradients are clipped in place in keeping with pytorch:
+// https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_norm_.html
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.hpp:16]: 👍

File: tt-train/tests/core/clip_grad_norm_test.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    serialization::NamedParameters named_params;
+    for (uint32_t i = 0; i < num_tensors; i++) {
+        named_params[std::to_string(i)] = tensors[i];
+    }
+    auto result = core::clip_grad_norm(named_params, max_norm, 2.0f);
+    float computed_total_norm = core::to_vector(result->get_value()).front();
+
+    // Verify total norm is approx. correct
+    // Total norm is a lot further off due to compounded bf16 errors
+    EXPECT_NEAR(computed_total_norm, expected_total_p2_norm, 1e-0f);
```
Comment (rfurko-tt) [inline tt-train/tests/core/clip_grad_norm_test.cpp:89]: 1e-0? is this expected?

File: tt-train/tests/core/clip_grad_norm_test.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    // Verify gradients were scaled approx. correctly
+    for (uint32_t i = 0; i < num_tensors; i++) {
+        auto grad_data = core::to_vector(tensors[i]->get_grad());
+        for (uint32_t j = 0; j < tensor_size; j++) {
+            EXPECT_NEAR(grad_data[j], expected_clipped_grads[i * tensor_size + j], 1e-1f);
```
Comment (rfurko-tt) [inline tt-train/tests/core/clip_grad_norm_test.cpp:95]: can we make 1e-1f smaller? 3e-2 as an example? 0.1 is larger than I expected

File: tt-train/sources/ttml/core/clip_grad_norm.cpp [hunk 5]
```diff
@@ -0,0 +1,26 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "core/clip_grad_norm.hpp"
+
+#include "core/ttnn_all_includes.hpp"
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.cpp:7]: substitute with #include <core/ttnn_all_includes.hpp>

File: tt-train/tests/core/clip_grad_norm_test.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    }
+};
+
+TEST_F(ClipGradNormTest, ClipGradNorm_GENEROUS_TOLERANCE) {
+    using namespace ttml;
+
+    auto* device = &autograd::ctx().get_device();
+
+    // Test tensors with manually set gradients
+    const uint32_t num_tensors = 3;
```
Comment (rfurko-tt) [inline tt-train/tests/core/clip_grad_norm_test.cpp:32]: const uint32_t num_tensors = 3U;

same with constant below

**Review by rfurko-tt — COMMENTED** [review]
Looks good overall, please also remove previous implementation and corresponding tests. Thanks in advance!

File: tt-train/tests/core/clip_grad_norm_test.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    void SetUp() override {
+        ttml::autograd::ctx().open_device();
+    }
+
+    void TearDown() override {
+        ttml::autograd::ctx().close_device();
+    }
+};
+
+TEST_F(ClipGradNormTest, ClipGradNorm_GENEROUS_TOLERANCE) {
```
Comment (dmakoviichuk-tt) [inline tt-train/tests/core/clip_grad_norm_test.cpp:26]: Nice name :-D but we have generous tolerance everywhere

File: tt-train/sources/ttml/core/clip_grad_norm.cpp [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+autograd::TensorPtr clip_grad_norm(
+    const serialization::NamedParameters& parameters, float max_norm, float p_norm_type, bool error_if_nonfinite) {
+    std::vector<tt::tt_metal::Tensor> grads;
+    grads.reserve(parameters.size());
+    for (const auto& [_, tensor] : parameters) {
+        if (tensor->get_requires_grad() && tensor->is_grad_initialized()) {
+            grads.push_back(tensor->get_grad());
+        }
+    }
+    auto tt_result = ttnn::moreh_clip_grad_norm(
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.cpp:21]: please check if function takes COmpute COnfig. and please use our core::ComputeKernelConfig::precise()

File: tt-train/tests/core/clip_grad_norm_test.cpp (hunk 3 above)
Comment (jaykru-tt) [reply tt-train/tests/core/clip_grad_norm_test.cpp:89]: yes, unfortunately the total norm can be very far off because it sums a bunch of terms with their own error.

File: tt-train/tests/core/clip_grad_norm_test.cpp (hunk 4 above)
Comment (jaykru-tt) [reply tt-train/tests/core/clip_grad_norm_test.cpp:95]: I'll play around a bit and see if I can tighten it up :) Denys also suggested that the precise compute config may allow us to tighten this even further.

File: tt-train/tests/core/clip_grad_norm_test.cpp (hunk 3 above)
Comment (rfurko-tt) [reply tt-train/tests/core/clip_grad_norm_test.cpp:89]: we should definitely try it out

Comment (jaykru-tt) [reply tt-train/tests/core/clip_grad_norm_test.cpp]: I tightened the bound a bit, but the results were the same even with the precise config 😢

Comment (jaykru-tt) [reply tt-train/tests/core/clip_grad_norm_test.cpp]: I'm seeing errors of .0625 elsewhere which strongly suggests this should go away after we whip up arith operators for metal's bfloat16 class and use that instead of f32.

File: tt-train/sources/ttml/core/clip_grad_norm.hpp [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+// Returns the summed norm of the gradients after clipping.
+autograd::TensorPtr clip_grad_norm(
+    const serialization::NamedParameters& parameters,
+    float max_norm,
+    float p_norm_type = 2.0f,
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.hpp:22]: f -> F

File: tt-train/sources/ttml/core/clip_grad_norm.cpp [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        grads,
+        max_norm,
+        p_norm_type,
+        error_if_nonfinite,
+        std::nullopt, /* total_norm*/
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.cpp:26]: add /* total_norm */ on the left of the value (style of the codebase)

Comment (rfurko-tt) [reply tt-train/sources/ttml/core/clip_grad_norm.cpp:26]: same for memory_config

File: tt-train/tests/core/clip_grad_norm_test.cpp [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        [](const xt::xarray<float>& acc, const xt::xarray<float>& curr) {
+            return xt::concatenate(xt::xtuple(acc, curr), 0);
+        });
+
+    float expected_total_p2_norm = p2_norm(concatenated_grads);
+
+    // reshape grads to 4D, required for compatibility with ttnn ops.
+    std::transform(
+        expected_grads.begin(), expected_grads.end(), expected_grads.begin(), [](const xt::xarray<float>& grad) {
+            return xt::xarray<float>(xt::reshape_view(grad, {1U, 1U, 1U, tensor_size}));
```
Comment (rfurko-tt) [inline tt-train/tests/core/clip_grad_norm_test.cpp:64]: can you just make inplace reshape?

File: tt-train/tests/core/clip_grad_norm_test.cpp [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    std::vector<xt::xarray<float>> expected_grads = {
+        {1.0F, 2.0F, 3.0F, 4.0F}, {-2.0F, -3.0F, -4.0F, 5.0F}, {0.5F, -1.5F, 2.5F, -3.5F}};
+
+    // Compute reference total p2 norm manually; the "total norm" in pytorch
+    // parlance is the L2 norm of tensor obtained by concatenating all the grads
+    // together.
+    auto p2_norm = [](const xt::xarray<float>& grad) {
+        float res = 0.0F;
+        for (auto it = grad.begin(); it != grad.end(); it++) {
+            res += std::pow(*it, 2);
```
Comment (rfurko-tt) [inline tt-train/tests/core/clip_grad_norm_test.cpp:46]: i think you can write it with xtensor vectorized operations sum(grad * grad), please correct me if i am wrong

File: tt-train/tests/core/clip_grad_norm_test.cpp (hunk 11 above)
Comment (rfurko-tt) [reply tt-train/tests/core/clip_grad_norm_test.cpp:64]: anyway copy is happening

File: tt-train/sources/ttml/core/clip_grad_norm.hpp (hunk 9 above)
Comment (jaykru-tt) [reply tt-train/sources/ttml/core/clip_grad_norm.hpp:22]: thanks, hopefully we can get clang-tidy to pick these up from now on 😆

File: tt-train/sources/ttml/core/clip_grad_norm.hpp [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    float max_norm,
+    float p_norm_type = 2.0F,
+    bool error_if_nonfinite = true);
+
+autograd::TensorPtr clip_grad_norm(const serialization::NamedParameters& parameters, const ClipGradNormConfig& config);
+
+}  // namespace ttml::core
+
+// for parsing from yaml model config
+namespace YAML {
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.hpp:38]: lets put it to the separate file. Also you cna move implementation to the c++ despite of that itit templatte

File: tt-train/configs/training_shakespear_nanogpt.yaml [hunk 14]
```diff
@@ -9,6 +9,10 @@ training_config:
   weight_decay: 0.01
   use_moreh_adamw: true
   use_kahan_summation: false
+  clip_grad_norm_config:
```
Comment (dmakoviichuk-tt) [inline tt-train/configs/training_shakespear_nanogpt.yaml:12]: I think only 1 parameter is enough: max_norm.
We will never change other.

File: tt-train/README.md [hunk 15]
```diff
@@ -53,7 +53,7 @@ cmake --build build --config Release --clean-first
 ### Training
 ```
 # Navigate to the root directory of the repository
-TT_METAL_LOGGER_LEVEL=FATAL ./build/sources/examples/nano_gpt/nano_gpt --model_path nano_gpt.msgpack --data_path sources/examples/nano_gpt/data/shakespeare.txt
+TT_METAL_LOGGER_LEVEL=FATAL ./build/sources/examples/nano_gpt/nano_gpt
```
Comment (dmakoviichuk-tt) [inline tt-train/README.md:56]: thanks for fixit docs :)

File: tt-train/sources/examples/nano_gpt/main.cpp [hunk 16]
```diff
@@ -622,6 +635,9 @@ int main(int argc, char **argv) {
             float loss_float = get_loss_value(loss);
 
             loss->backward();
+            if (config.clip_grad_norm_config.has_value()) {
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/nano_gpt/main.cpp:638]: @rfurko-tt  please double check two cases:
gradient_accumulator and ddp. I am not sure it ithe right place for both cases.

File: tt-train/sources/ttml/core/clip_grad_norm.hpp [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+#pragma once
+#include <yaml-cpp/yaml.h>
+
+#include "autograd/tensor.hpp"
+#include "serialization/serializable.hpp"
+
+namespace ttml::core {
+
+struct ClipGradNormConfig {
+    float max_norm;
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.hpp:14]: don't forget to always make default initialization.

File: tt-train/sources/examples/nano_gpt/main.cpp (hunk 16 above)
Comment (rfurko-tt) [reply tt-train/sources/examples/nano_gpt/main.cpp:638]: discussed offline.
1. no need to adjust for gradient accumulation
2. remove gradient sync from optimizer, add explicit calls in examples

File: tt-train/sources/ttml/core/clip_grad_norm.hpp (hunk 13 above)
Comment (jaykru-tt) [reply tt-train/sources/ttml/core/clip_grad_norm.hpp]: Removed this entirely

File: tt-train/sources/ttml/core/clip_grad_norm.hpp (hunk 17 above)
Comment (jaykru-tt) [reply tt-train/sources/ttml/core/clip_grad_norm.hpp]: Removed this, but noted for the future :)

File: tt-train/configs/training_shakespear_nanogpt.yaml [hunk 18]
```diff
@@ -9,6 +9,8 @@ training_config:
   weight_decay: 0.01
   use_moreh_adamw: true
   use_kahan_summation: false
+  use_clip_grad_norm: true
```
Comment (rfurko-tt) [inline tt-train/configs/training_shakespear_nanogpt.yaml:12]: please disable by default

File: tt-train/sources/examples/nano_gpt/main.cpp [hunk 19]
```diff
@@ -627,6 +635,12 @@ int main(int argc, char **argv) {
             auto samples = features->get_value().get_shape()[0];
             gradient_accumulator_helper.update(loss_float, samples);
 
+            // synchronize gradients for multi-device case, no-op if single device
+            ttml::core::distributed::synchronize_parameters(model->parameters());
```
Comment (rfurko-tt) [inline tt-train/sources/examples/nano_gpt/main.cpp:639]: could you please request model->parameters once? 
```
auto model_params = model->parameters();
ttml::core::distributed::synchronize_parameters(model_params);
if (...) {
    ttml::core::clip_grad_norm(model_params, config.clip_grad_norm_max_norm);
}
```

File: tt-train/sources/ttml/core/clip_grad_norm.cpp (hunk 10 above)
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.cpp:26]: move comment to the left

File: tt-train/sources/ttml/core/clip_grad_norm.hpp [hunk 20]
```diff
@@ -0,0 +1,25 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include <yaml-cpp/yaml.h>
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/core/clip_grad_norm.hpp:6]: remove yaml header

File: tt-train/tests/core/clip_grad_norm_test.cpp [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    std::vector<autograd::TensorPtr> tensors;
+    std::vector<xt::xarray<float>> expected_grads = {
+        {1.0F, 2.0F, 3.0F, 4.0F}, {-2.0F, -3.0F, -4.0F, 5.0F}, {0.5F, -1.5F, 2.5F, -3.5F}};
+
+    // Compute reference total p2 norm manually; the "total norm" in pytorch
+    // parlance is the L2 norm of tensor obtained by concatenating all the grads
+    // together.
+    auto p2_norm = [](const xt::xarray<float>& grad) { return std::sqrt(xt::sum(grad * grad)()); };
+
+    auto concatenated_grads = std::accumulate(
```
Comment (rfurko-tt) [inline tt-train/tests/core/clip_grad_norm_test.cpp:45]: use util function for concatenate

**Review by rfurko-tt — COMMENTED** [review]
Thanks for the fixes! 
Almost there :) 
Please add charts from w&b with main vs new vs new_with_clip_grad_enabled.

**Comment by jaykru-tt** [comment]
Thanks for the review @rfurko-tt :) Will add charts shortly.

**Comment by jaykru-tt** [comment]
image
Let me know if there's a better way to share chart comparisons than this.
Key: 
- Black: main/baseline
- Red: with change, NOT using clip grad norm: identical to main.
- Green: with changes, using clip grad norm: moderate improvement in loss curve

File: tt-train/sources/ttml/core/distributed.hpp [hunk 22]
```diff
@@ -0,0 +1,14 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "serialization/serializable.hpp"
+
+namespace ttml::core::distributed {
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/core/distributed.hpp:9]: in tt-train our namespaces are always equal to the folders. Could you create distributed folder please?

Comment (jaykru-tt) [reply tt-train/sources/ttml/core/distributed.hpp]: done :)

**Review by rfurko-tt — APPROVED** [review]
LGTM, great job!

**Comment by rfurko-tt** [comment]
@jaykru-tt add charts to PR description, not comment :)

**Review by dmakoviichuk-tt — APPROVED** [review]
